### DISCUSSION TITLE
Fix operator overloading to allow use of built in operator in overload

### DISCUSSION THREAD

**Discussion by JulianKemmerer** [JulianKemmerer/PipelineC#341] [Ideas]
Operator overloading can't use built in version of operator in definition of overload. Fix this to separate user's version of built in vs built in...

Work around is to use the generated C code from the built in version as user code under different function name...

**Comment by RossBencina** [comment]
One way you might be able to do this would be to define all builtin operator definitions as functions with a `BUILTIN_` prefix, and `#define` them to the built-in operator names, equivalent to:

```
float BUILTIN_BIN_OP_PLUS_float_float(float left, float right)
{
... builtin implementation goes here, i know this def might be hardcoded elsewhere ...
}

#define BIN_OP_PLUS_float_float(left, right) BUILTIN_BIN_OP_PLUS_float_float(left, right)
```

Then if the user wants to replace it with their own:

```
#undef BIN_OP_PLUS_float_float
float BIN_OP_PLUS_float_float(float left, float right)
{
    ... user definition ...
}
```

Another way would just be to use nested symbol tables and do operator name lookup first in the user-defined functions symbol table and only fall back to the global/builtin symbol table if the user has not provided an override.

**Comment by JulianKemmerer** [comment]
Certainly sounds plausible to me I have to think a little more about what you are suggesting 

What happens now is for reference:
Upon reaching an operator, and determining its types, the identifier of ex. `BIN_OP_PLUS_float_float` is looked up in the symbol table. If that func name is provided by the user in code then that version of implementation is used. Otherwise, python code gen for a C implementation of that operator for those types is done, and that code is parsed, added to symbol table (just as if user had written it)

**Comment by JulianKemmerer** [comment]
Moving to Discussions: this is a feature idea rather than something broken in a design today. Especially if C front end effected only.
